src/ripple/overlay/impl/OverlayImpl.cpp

```diff
@@ -237,88 +237,84 @@ OverlayImpl::onHandoff (std::unique_ptr <beast::asio::ssl_bundle>&& ssl_bundle,
         }
     }
 
-    auto hello = parseHello (true, request, journal);
-    if(! hello)
+    auto negotiatedVersion = negotiateProtocolVersion(request["Upgrade"]);
+    if (!negotiatedVersion)
     {
         m_peerFinder->on_closed(slot);
         handoff.moved = false;
         handoff.response = makeErrorResponse (slot, request,
-            remote_endpoint.address(),
-            "Unable to parse HELLO message");
+            remote_endpoint.address(), "Unable to agree on a protocol version");
         handoff.keep_alive = false;
         return handoff;
     }
 
-    auto sharedValue = makeSharedValue(
-        ssl_bundle->stream.native_handle(), journal);
+    auto sharedValue = makeSharedValue(*ssl_bundle, journal);
     if(! sharedValue)
     {
         m_peerFinder->on_closed(slot);
         handoff.moved = false;
         handoff.response = makeErrorResponse (slot, request,
-            remote_endpoint.address(),
-            "Incorrect security cookie (possible MITM detected)");
+            remote_endpoint.address(), "Incorrect security cookie");
         handoff.keep_alive = false;
         return handoff;
     }
 
-    auto publicKey = verifyHello (*hello,
-        *sharedValue,
-        setup_.public_ip,
-        beast::IPAddressConversion::from_asio(
-            remote_endpoint), journal, app_);
-    if(! publicKey)
+    try
```

Comment:
I believe this `try` block is based on the fact that `verifyHandshake` now throws. Since it has a much larger scope than just the `verifyHandshake` call, once we add it's much more difficult to remove it since we have to audit all the calls in the scope for throws.

Reducing the scope of the `try` makes the code uglier, I'm not advocating that; I am pointing out that once we decide `verifyHandshake` throws it's hard to reverse that decision.